### PR TITLE
Added CSI-2 driver implementation docs

### PR DESCRIPTION
**Pull request by JamesH65**

### CODE REVIEW

**Comment by JamesH65** [comment]
This will still need some copy edit input I suspect, I also suspect they are busy.

**Comment by JamesH65** [comment]
I should add that this doc was written almost entirely by @6by9, I take no credit for it at all.

File: linux/software/libcamera/csi-2-usage.md [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+
+### Bridge chips
+
+These are devices that convert an incoming video stream, for example HDMI or composite, into a CSI-2 stream that can be accepted by the Raspberry Pi CSI-2 receiver.
+
+Handling bridge chips is more complicated as unlike camera sensors they have to respond to the incoming signal and report that to the application. For example, with a camera sensor the application tells it what resolution to output. With a incoming video stream it is the stream itself that provides the resolution that must be passed on to the application. 
+
+The mechanisms for handling bridge chips can be broadly split into either analogue or digital. 
+
+When using `ioctls` in the sections below, an `_S_` in the `ioctl` name means it is a set function, whilst `_G_` is a get function.
```
Comment (6by9) [inline linux/software/libcamera/csi-2-usage.md:98]: _ENUM for enumerating permitted values.

File: linux/software/libcamera/csi-2-usage.md [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+ Selecting the wrong standard will generally result in corrupt images. Setting the standard will typically also set the resolution on the V4L2 CAPTURE queue. It can not be set via `VIDIOC_S_FMT`. Generally requesting the detected standard via `VIDIOC_QUERYSTD` and then setting it with `VIDIOC_S_STD` before streaming is a good idea. 
+
+#### Digital video sources
+
+For digital video sources, such as HDMI, there is an alternate set of calls that allow specifying of all the digital timing parameters ([`VIDIOC_G_DV_TIMINGS`](https://www.kernel.org/doc/html/latest/media/uapi/v4l/vidioc-g-dv-timings.html), [`VIDIOC_S_DV_TIMINGS`](https://www.kernel.org/doc/html/latest/media/uapi/v4l/vidioc-g-dv-timings.html), [`VIDIOC_ENUM_DV_TIMINGS`](https://www.kernel.org/doc/html/latest/media/uapi/v4l/vidioc-enum-dv-timings.html), and [`VIDIOC_QUERY_DV_TIMINGS`](https://www.kernel.org/doc/html/latest/media/uapi/v4l/vidioc-query-dv-timings.html) - links again).
```
Comment (6by9) [inline linux/software/libcamera/csi-2-usage.md:108]: "links again" I think was a note to add them, which you've done.

File: linux/software/libcamera/csi-2-usage.md [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+*Analog Devices ADV728x(A)-M Analogue video to CSI2 bridge*
+
+These chips convert composite, S-video (Y/C), or component (YPrPb) video into a single lane CSI-2 interface, and are supported by the [ADV7180 kernel driver](https://github.com/raspberrypi/linux/blob/rpi-5.4.y/drivers/media/i2c/adv7180.c).
+
+Product details for the various versions of this chip can be found on the Analog Devices website. 
+
+[ADV7280A](https://www.analog.com/en/products/adv7280a.html), [ADV7281A](https://www.analog.com/en/products/adv7281a.html), [ADV7282A](https://www.analog.com/en/products/adv7282a.html)
+
+Because of some missing code in the current core V4L2 implementation, selecting the source fails, so the Raspberry Pi kerrnel version adds a kernel module parameter called `dbg_input` to the ADV7180 kernel driver which sets the input source every time VIDIOC_S_STD is called. At some point mainstream will fix the underlying issue (a disjoin between the kernel API call s_routing, and the userspace call `VIDIOC_S_INPUT`) and this modification will be removed.
```
Comment (6by9) [inline linux/software/libcamera/csi-2-usage.md:127]: s/kerrnel/kernel

File: linux/software/libcamera/csi-2-usage.md [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+
+Please note that receiving interlaced video is not supported, therefore the ADV7281(A)-M version of the chip is of limited use as it doesn't have the necessary I2P deinterlacing block. Also ensure when selecting a device to specify the -M option. Without that you will get a parallel output bus which can not be interfaced to the Raspberry Pi.
+
+There are no known commercially available boards using these chips, but this driver has been tested via the Analog Devices [EVAL-ADV7282-M evaluation board](https://www.analog.com/en/design-center/evaluation-hardware-and-software/evaluation-boards-kits/EVAL-ADV7282A-M.html)
+
+This driver can be loaded using the `config.txt` dtoverlay entry `adv7282m`, or `adv728x` with a parameter of either `adv7280m=1`,`adv7281m=1`, or `adv7281ma=1` if you are not using the ADV7282-M chip variant. e.g.
```
Comment (6by9) [inline linux/software/libcamera/csi-2-usage.md:133]: drop the word "entry"

File: linux/software/libcamera/csi-2-usage.md [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+```
+dtoverlay=adv728x=adv7280m=1
```
Comment (6by9) [inline linux/software/libcamera/csi-2-usage.md:136]: dtoverlay=adv728x,adv7280m=1
(comma not equals)

File: linux/software/libcamera/csi-2-usage.md (hunk 4 above)
Comment (6by9) [reply linux/software/libcamera/csi-2-usage.md:133]: And I've got the overlay name wrong - it's adv728x-m

File: linux/software/libcamera/csi-2-usage.md [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+```
+
+*Toshiba TC358743 HDMI to CSI2 bridge*
+
+This is a HDM I to CSI-2 bridge chip, capable of converting video data at up to 1080p60. Although capable of 4 lane transfers, on the Raspberry only 2 lanes are used.
```
Comment (6by9) [inline linux/software/libcamera/csi-2-usage.md]: "HDM I"

Comment (6by9) [reply linux/software/libcamera/csi-2-usage.md]: Drop the second sentence - it's covered later.
(4 lanes can be used on the CM/CM3, therefore it's also ambiguous/wrong)

**Comment by JamesH65** [comment]
I've taken off the Ready for copy edit to wait for any other input from reviewers, once they are happy, it will go to CE, but that take about a week.

File: linux/software/libcamera/csi-2-usage.md [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+### MMAL rawcam component
+
+This was an interim option before the V4L2 driver was available. The MMAL component `vc.ril.rawcam` allows receiving of the raw CSI2 data in the same way as the V4L2 driver, but all source configuration has to be done by userland over whatever interface the source requires. The raspiraw application is available on [github]( https://github.com/raspberrypi/raspiraw). It uses this component and the standard I2C register sets for OV5647, IMX219, and ADV7282M to support streaming.
+
+
+### V4L2
+
+There is a fully open source kernel driver available for the Unicam block, this is a kernel module called bcm2835-unicam. This interfaces to V4L2 subdevice drivers for the source to deliver the raw frames. This bcm2835-unicam driver controls the sensor, and configures the CSI-2 receiver so that the peripheral will write the raw frames (after Debayer) to SDRAM for V4L2 to deliver to applications. Except for this ability to unpack the CSI-2 Bayer formats to 16bits/pixel, there is no image processing between the image source (e.g. camera sensor) and brcm2835-unicam placing the image data in SDRAM.
```
Comment (andrum99) [inline linux/software/libcamera/csi-2-usage.md:25]: Comma splice - Unicam block, this is -> Unicam block; this is

File: linux/software/libcamera/csi-2-usage.md [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+| Omnivision OV5647 | 5MP Camera | Original Raspberry Pi Camera |
+| Sony IMX219 | 8MP Camera | Revision 2 Raspberry Pi camera |
+| Sony IMX477 | 12MP Camera | Raspberry Pi HQ camera |
+| Toshiba TC358743 | HDMI to CSI-2 bridge | |
+| Analog Devices ADV728x-M | Analog video to CSI-2 bridge| No interlaced support |
+| Infineon IRS1125 | Time-of-flight depth sensor| Supported by a third party |
+
+As the subdevice driver is also a kernel driver, with a standardised API, 3rd parties are free to write their own for any source of their choosing. 
+
+## Developing a Third-Party driver for bcm2835-unicam
```
Comment (andrum99) [inline linux/software/libcamera/csi-2-usage.md:59]: Third-Party -> third-party

File: linux/software/libcamera/csi-2-usage.md [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+
+This is the recommended approach to interfacing via Unicam.
+
+When developing a driver for a new device intended to be used with the bcm2835-unicam module, you need the driver and corresponding device tree overlays. Ideally the driver should be submitted to the linux-media mailing list for code review and merging into mainline, then moved to the Raspberry Pi kernel tree, but exceptions may be made for the driver to be reviewed and merged directly to the Raspberry Pi kernel.
+
+Please note that the all kernel drivers are licenced under the GPLv2 licence, therefore source code **MUST** be available. Shipping of binary modules only is a violation of the GPLv2 licence under which the Linux kernel is licenced.
```
Comment (andrum99) [inline linux/software/libcamera/csi-2-usage.md:65]: the all -> all

File: linux/software/libcamera/csi-2-usage.md [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+
+- `V4L2_CID_PIXEL_RATE` / `V4L2_CID_VBLANK` / `V4L2_CID_HBLANK`: allows the application to set the frame rate.
+- `V4L2_CID_EXPOSURE`: sets the exposure time in lines. The application needs to use `V4L2_CID_PIXEL_RATE`, `V4L2_CID_HBLANK`, and the frame width to compute the line time.
+- `V4L2_CID_ANALOGUE_GAIN`: analogue gain in sensor specific units.
+- `V4L2_CID_DIGITAL_GAIN`: optional digital gain in sensor specific units.
+- `V4L2_CID_HFLIP / V4L2_CID_VFLIP`: flips the image either horizontally or vertically. Note that this operation may change the Bayer order of the data in the frame, as is the case on the imx219.
+- `V4L2_CID_TEST_PATTERN` / `V4L2_CID_TEST_PATTERN_*`: Enables output of various test patterns from the sensor. Useful for debugging.
+
+In the case of the IMX219, many of these controls map directly on to register writes to the sensor itself.
```
Comment (andrum99) [inline linux/software/libcamera/csi-2-usage.md:84]: on to -> onto

File: linux/software/libcamera/csi-2-usage.md [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+| Analog Devices ADV728x-M | Analog video to CSI-2 bridge| No interlaced support |
+| Infineon IRS1125 | Time-of-flight depth sensor| Supported by a third party |
+
+As the subdevice driver is also a kernel driver, with a standardised API, 3rd parties are free to write their own for any source of their choosing. 
+
+## Developing a Third-Party driver for bcm2835-unicam
+
+This is the recommended approach to interfacing via Unicam.
+
+When developing a driver for a new device intended to be used with the bcm2835-unicam module, you need the driver and corresponding device tree overlays. Ideally the driver should be submitted to the linux-media mailing list for code review and merging into mainline, then moved to the Raspberry Pi kernel tree, but exceptions may be made for the driver to be reviewed and merged directly to the Raspberry Pi kernel.
```
Comment (andrum99) [inline linux/software/libcamera/csi-2-usage.md:63]: backticks round linux-media?

Comment (JamesH65) [reply linux/software/libcamera/csi-2-usage.md:63]: Actually made it a link.

File: linux/software/libcamera/csi-2-usage.md [hunk 12]
```diff
@@ -56,13 +56,13 @@ Mainline Linux has a range of existing drivers. The Raspberry Pi kernel tree has
 
 As the subdevice driver is also a kernel driver, with a standardised API, 3rd parties are free to write their own for any source of their choosing. 
 
-## Developing a Third-Party driver for bcm2835-unicam
+## Developing a third-Party driver for bcm2835-unicam
```
Comment (andrum99) [inline linux/software/libcamera/csi-2-usage.md:59]: Party -> party 😶

Comment (JamesH65) [reply linux/software/libcamera/csi-2-usage.md:59]: DOH!

File: linux/software/libcamera/csi-2-usage.md [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+
+### MMAL rawcam component
+
+This was an interim option before the V4L2 driver was available. The MMAL component `vc.ril.rawcam` allows receiving of the raw CSI2 data in the same way as the V4L2 driver, but all source configuration has to be done by userland over whatever interface the source requires. The raspiraw application is available on [github]( https://github.com/raspberrypi/raspiraw). It uses this component and the standard I2C register sets for OV5647, IMX219, and ADV7282M to support streaming.
+
+
+### V4L2
+
+There is a fully open source kernel driver available for the Unicam block; this is a kernel module called bcm2835-unicam. This interfaces to V4L2 subdevice drivers for the source to deliver the raw frames. This bcm2835-unicam driver controls the sensor, and configures the CSI-2 receiver so that the peripheral will write the raw frames (after Debayer) to SDRAM for V4L2 to deliver to applications. Except for this ability to unpack the CSI-2 Bayer formats to 16bits/pixel, there is no image processing between the image source (e.g. camera sensor) and brcm2835-unicam placing the image data in SDRAM.
```
[the remark this answers is not in the document]
Comment (6by9) [reply linux/software/libcamera/csi-2-usage.md:25]: Good spot. It's bcm2835-unicam in all cases.

**Comment by JamesH65** [comment]
OK, sending this off for copy edit now, think most of what we want is in there.

File: linux/software/libcamera/csi-2-usage.md [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+
+As the subdevice driver is also a kernel driver, with a standardised API, 3rd parties are free to write their own for any source of their choosing. 
+
+## Developing a third-party driver for bcm2835-unicam
+
+This is the recommended approach to interfacing via Unicam.
+
+When developing a driver for a new device intended to be used with the bcm2835-unicam module, you need the driver and corresponding device tree overlays. Ideally the driver should be submitted to the [linux-media](http://vger.kernel.org/vger-lists.html#linux-media) mailing list for code review and merging into mainline, then moved to the [Raspberry Pi kernel tree](https://github.com/raspberrypi/linux), but exceptions may be made for the driver to be reviewed and merged directly to the Raspberry Pi kernel.
+
+Please note that all kernel drivers are licensed under the GPLv2 license, therefore source code **MUST** be available. Shipping of binary modules only is a violation of the GPLv2 license under which the Linux kernel is licensed.
```
Comment (lurch) [inline linux/software/libcamera/csi-2-usage.md:64]: license -> licence :wink: 

(it's spelt with 's' when used as a verb, but spelt with 'c' when used as a noun. E.g. TV Licences are issued by the TV Licensing Authority.)

Comment (JamesH65) [reply linux/software/libcamera/csi-2-usage.md:64]: I probably knew that, once.

File: linux/software/libcamera/csi-2-usage.md [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+This driver is loaded using the `config.txt` dtoverlay `tc358743`.
+
+The chip also supports capturing stereo HDMI audio via I2S. The Auvidea boards break the relevant signals out onto a header, which can be connected to the Pi's 40 pin header. The required wiring is:
+
+| Signal   | B101 header | Pi 40 pin header |
+|----------|:-----------:|:----------------:|
+| LRCK/WFS |     7       |      19          |
+| BCK/SCK  |     6       |      18          |
+| DATA/SD  |     5       |      20          |
+| GND      |     8       |      39          |
```
Comment (6by9) [inline linux/software/libcamera/csi-2-usage.md:163]: Oops, my bad here. Those are BCM GPIO numbers, not 40pin header pin numbers. Another column needed
```
| Signal   | B101 header | Pi 40 pin header | BCM GPIO  |
|----------|:-----------:|:----------------:|:---------:|
| LRCK/WFS |     7       |       35         |    19     |
| BCK/SCK  |     6       |       12         |    18     |
| DATA/SD  |     5       |       38         |    20     |
| GND      |     8       |       39         |    N/A    |

**Comment by 6by9** [comment]
Gentle ping for copy edit after 2 months.
This documentation is important for third party libcamera support, and it's slightly embarrassing having to link people to a PR instead of the docs.

**Pull request merged**